# Correct Resolution Models

The following models take into account that, throughout analyses, *no_goal* seems to behave entirely differently to the other two Goal Type conditions (*goal_frequent* and *goal_non_frequent*). 

Previous focus structure effects (see notebooks 11 to 6) as well as further exhaustive modelling (see notebooks 15_appendix vs. 16_appendix) show consistently the heterogeneity of combined data; by spliting data into goal contexts and no-goal contexts, both effects as well as models improved despite the pruning of observations. No-goal contexts are also analysed separately ((notebook 11)); however, and although results are coherent with what is seen by comparing combined data to goal behaviour data, the sample is clearly not enough to show anything significant from no-goal contexts.

Taking into account the effects summary (notebook 5) and response opportunity analysis (notebook 6), these were chosen to be the best models to predict participants resolving succesfully despite lack of significance and adding predictors supported by opportunity analysis. However, unlike Escape_L2 and Transfer, correct resolution is only partially explained by the predictors available in this dataset. 

Import Libraries

In [116]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Goal Behaviour Subgroup

1. Create resolved_correct
2. Counts and Proportions
3. Logistic Models

     Main Effect Models:

    - RC01: Focus

    Additive Models:

    - RC05: Focus + Goal_Type

    Interactive Models:

    - RC11: Focus * Agent + Goal_Type

4. Summary

Read Data

In [117]:
subgroup_theoretical = pd.read_csv("../../data/processed/subgroup_theoretical.csv")

1. Create resolved_correct

Binary Outcome: correct responses from escape_L2 observations.

In [118]:
subgroup_theoretical["escape_L2"] = (
    subgroup_theoretical["Response_Full"] != "L2_other"
).astype(int)

In [119]:
escapees = subgroup_theoretical[subgroup_theoretical["escape_L2"] == 1].copy()

In [120]:
escapees["resolved_correct"] = (escapees["Response_Full"] == "correct").astype(int)

Split escapees data into goals vs. no_goals

In [121]:
goals_escapees = escapees[escapees["Goal_Type"] != "no_goal"].copy()

In [122]:
no_goal_escapees = escapees[escapees["Goal_Type"] == "no_goal"].copy()

2. Counts and Proportions for goals

resolved_correct responses count for goals:

In [123]:
goals_resolved_correct_counts = pd.crosstab(
    goals_escapees["resolved_correct"],
    goals_escapees["Response_Full"],
    margins= True)

goals_resolved_correct_counts

Response_Full,L1_transfer,correct,missing_response,All
resolved_correct,,,,
0,62,0,1,63
1,0,43,0,43
All,62,43,1,106


resolved_correct responses proportions for goals:

In [124]:
goals_resolved_correct_props = goals_escapees["resolved_correct"].value_counts(normalize=True)
goals_resolved_correct_props

resolved_correct
0    0.59434
1    0.40566
Name: proportion, dtype: float64

From 106 observations, 40.5% of goal responses that had escaped *L2_other* options were resolved correctly. 

Counts and Proportions by Condition:

In [125]:
goals_counts_condition = pd.crosstab(
    [goals_escapees["Goal_Type"], goals_escapees["Agent"], goals_escapees["Focus"]],
    goals_escapees["resolved_correct"],
    margins = True)

goals_counts_condition

resolved_correct                0   1  All
Goal_Type         Agent Focus             
goal_frequent     0     I       7   9   16
                        They    7   5   12
                  1     I       7   7   14
                        They    5   3    8
goal_non_frequent 0     I       7   9   16
                        They   14   3   17
                  1     I      11   4   15
                        They    5   3    8
All                            63  43  106

In [126]:
goals_props_condition = pd.crosstab(
    [goals_escapees["Goal_Type"], goals_escapees["Agent"], goals_escapees["Focus"]],
    goals_escapees["resolved_correct"],
    normalize= "index")

goals_props_condition

resolved_correct                      0         1
Goal_Type         Agent Focus                    
goal_frequent     0     I      0.437500  0.562500
                        They   0.583333  0.416667
                  1     I      0.500000  0.500000
                        They   0.625000  0.375000
goal_non_frequent 0     I      0.437500  0.562500
                        They   0.823529  0.176471
                  1     I      0.733333  0.266667
                        They   0.625000  0.375000

3. Logistic Models

Import Libraries

In [127]:
import statsmodels.formula.api as smf
from scipy.stats import chi2

Sanity check:

In [128]:
goals_escapees["resolved_correct"].value_counts()

resolved_correct
0    63
1    43
Name: count, dtype: int64

**Main Effect Models**

**RC01**

resolved_correct ~ Focus

In [129]:
RC01 = smf.logit(
    "resolved_correct ~ Focus",
    data=goals_escapees
    ).fit()

print(RC01.summary())

Optimization terminated successfully.
         Current function value: 0.661392
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  106
Model:                          Logit   Df Residuals:                      104
Method:                           MLE   Df Model:                            1
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.02051
Time:                        20:18:38   Log-Likelihood:                -70.108
converged:                       True   LL-Null:                       -71.575
Covariance Type:            nonrobust   LLR p-value:                   0.08664
                    coef    std err          z      P>|z|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept        -0.0984      0.256     -0.384      0.701      -0.601       0.404
Focus[T.They]    -0.

Focus, again, gains more relevance again when it comes to resolution. The model, however, is still not significant.

This clearly differs from escaping L2 other options which was clearly driven by Agency (LLR p-value = 0.001628) and, as expected, shows a similar behaviour to resolving by means of transfer which was also siding towards Focus (LLR p-value= 0.06029) as opposed to Agency. 

**Additive Model**

**RC05**

resolved_correct ~ Focus + Goal_Type

In [130]:
RC05 = smf.logit(
    "resolved_correct ~ Focus + Goal_Type",
    data=goals_escapees
    ).fit()

print(RC05.summary())

Optimization terminated successfully.
         Current function value: 0.651957
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  106
Model:                          Logit   Df Residuals:                      103
Method:                           MLE   Df Model:                            2
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.03448
Time:                        20:18:38   Log-Likelihood:                -69.107
converged:                       True   LL-Null:                       -71.575
Covariance Type:            nonrobust   LLR p-value:                   0.08476
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.1894      0.329      0.576     

Focus alone's LLR p-value = 0.0866; together, LLR p-value = 0.0848. Approaching significance, not there yet, but slightly better than the model with Focus on its own.

**Interactive Model**

**RC11**

resolved_correct ~ Focus * Agent + Goal_Type

In [131]:
RC11 = smf.logit(
    "resolved_correct ~ Focus * Agent + Goal_Type",
    data=goals_escapees
    ).fit()

print(RC11.summary())

Optimization terminated successfully.
         Current function value: 0.640518
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:       resolved_correct   No. Observations:                  106
Model:                          Logit   Df Residuals:                      101
Method:                           MLE   Df Model:                            4
Date:                Fri, 07 Aug 2026   Pseudo R-squ.:                 0.05142
Time:                        20:18:38   Log-Likelihood:                -67.895
converged:                       True   LL-Null:                       -71.575
Covariance Type:            nonrobust   LLR p-value:                    0.1180
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
Intercept                          0.5378      0.418      1.287     

- This full model is run to be compared to transfer resolution (LLR p-value: 0.05318) Focus p-value reaches 0.031. The Goal type coefficient is very similar to the previous model RC05 with focus and goal type (-0.5703), so its contribution is steady even though it doesn't improve its p-value. The interaction doesn't reach significance. 

- However, although RC05 was retained as the principal additive model because of its simplicity and comparable overall fit, the higher-order model RC11's coefficients mirrored those obtained for transfer resolution (RT11), with opposite signs across predictors. This pattern is consistent with the interpretation that transfer and correct production represent competing destinations for the same pool of responses released following L2 suppression.

4. Summary 


- For escaping L2, Agency woke up and Focus helped once Agency was present. Agency seems to affect whether the opportunity exists. 

- Once the data has been isolated by goal behaviour and considering both tranfer and succesful resolution, Focus becomes the dominant story, not Agency.  Together with goal type, they seem to affect what happens once the opportunity exists. Correct resolution wins less often.

- For correct resolution (as for transfer resolution), Focus becomes stronger. This shows that there could be different cognitive processes involved in leaving the L2-other space compared to succesful resolution.

- In comparison:

    - RC01: Focus explains most of what can be explained.
    - RC05: Goal Type adds a consistent but weak distribution.
    - RC11: Mirrors RT11 structurally but lacks sufficient power for stable estimation.